src/datajudge/constraints/base.py

```diff
@@ -25,11 +28,31 @@ def uncommon_substrings(string1: str, string2: str) -> Tuple[str, str]:
 @dataclass(frozen=True)
 class TestResult:
     outcome: bool
-    failure_message: Optional[str] = field(default=None, repr=False)
-    constraint_description: Optional[str] = field(default=None, repr=False)
+    _failure_message: Optional[str] = field(default=None, repr=False)
+    _constraint_description: Optional[str] = field(default=None, repr=False)
     _factual_queries: Optional[str] = field(default=None, repr=False)
     _target_queries: Optional[str] = field(default=None, repr=False)
 
+    def formatted_failure_message(self, formatter: Formatter) -> Optional[str]:
+        return (
+            formatter.fmt_str(self._failure_message) if self._failure_message else None
+        )
+
+    def formatted_constraint_description(self, formatter: Formatter) -> Optional[str]:
+        return (
+            formatter.fmt_str(self._constraint_description)
+            if self._constraint_description
+            else None
+        )
+
+    @property
+    def failure_message(self) -> Optional[str]:
```

Comment:
I like how you are using properties as one of the cookbook intentions, to add behavior to an instance attribute.